File: Classes and Functions/Python/AESCipher.py
"""
`pip install cryptography`

"""

from cryptography.hazmat.primitives import hashes, padding
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.backends import default_backend
import os
import base64

class AESCipher:
    def __init__(self, key: bytes, salt: bytes, iterations: int = 100_000):
        """
        Initialize the cipher with a password-based key derivation using SHA256.
        :param key: The user-provided key (e.g., a password or secret).
        :param salt: A salt for KDF (must be securely generated and saved).
        :param iterations: Number of iterations for key derivation.
        """
        self.backend = default_backend()
        self.key = self._derive_key(key, salt, iterations)

    def _derive_key(self, password: bytes, salt: bytes, iterations: int) -> bytes:
        kdf = PBKDF2HMAC(
            algorithm=hashes.SHA256(),
            length=32,  # AES-256 requires a 32-byte key
            salt=salt,
            iterations=iterations,
            backend=self.backend
        )
        return kdf.derive(password)

    def encrypt(self, plaintext: bytes) -> bytes:
        iv = os.urandom(16)
        padder = padding.PKCS7(128).padder()
        padded_data = padder.update(plaintext) + padder.finalize()

        cipher = Cipher(algorithms.AES(self.key), modes.CBC(iv), backend=self.backend)
        encryptor = cipher.encryptor()
        ciphertext = encryptor.update(padded_data) + encryptor.finalize()

        return base64.b64encode(iv + ciphertext)

    def decrypt(self, encrypted_data: bytes) -> bytes:
        try:
            raw = base64.b64decode(encrypted_data)
            if len(raw) < 16:
                raise ValueError("Invalid encrypted data: too short for IV")

            iv = raw[:16]
            ciphertext = raw[16:]

            cipher = Cipher(algorithms.AES(self.key), modes.CBC(iv), backend=self.backend)
            decryptor = cipher.decryptor()
            padded_plaintext = decryptor.update(ciphertext) + decryptor.finalize()

            unpadder = padding.PKCS7(128).unpadder()
            plaintext = unpadder.update(padded_plaintext) + unpadder.finalize()
            return plaintext

        except Exception as e:
            raise ValueError(f"Decryption failed: {str(e)}")

# Example usage
if __name__ == "__main__":
    key = b'08180230818209389012832132109123'
    salt = os.urandom(16)

    aes = AESCipher(key, salt)

    data = b'This is very secret data!'
    encrypted = aes.encrypt(data)
    print("Encrypted:", encrypted)

    decrypted = aes.decrypt(encrypted)
    print("Decrypted:", decrypted)

File: Classes and Functions/Javascript/AESCipher.js
// AESCipher.js
const crypto = require('crypto');

class AESCipher {
  constructor(password, salt, iterations = 100000) {
    this.key = this.deriveKey(password, salt, iterations);
  }

  deriveKey(password, salt, iterations) {
    return crypto.pbkdf2Sync(password, salt, iterations, 32, 'sha256');
  }

  encrypt(plaintext) {
    const iv = crypto.randomBytes(16);
    const cipher = crypto.createCipheriv('aes-256-cbc', this.key, iv);

    const padded = this.pkcs7Pad(Buffer.from(plaintext));
    const ciphertext = Buffer.concat([cipher.update(padded), cipher.final()]);

    return Buffer.concat([iv, ciphertext]).toString('base64');
  }

  decrypt(encryptedBase64) {
    const raw = Buffer.from(encryptedBase64, 'base64');
    if (raw.length < 16) throw new Error('Invalid encrypted data');

    const iv = raw.slice(0, 16);
    const ciphertext = raw.slice(16);

    const decipher = crypto.createDecipheriv('aes-256-cbc', this.key, iv);
    const paddedPlaintext = Buffer.concat([decipher.update(ciphertext), decipher.final()]);
    return this.pkcs7Unpad(paddedPlaintext).toString();
  }

  pkcs7Pad(buffer) {
    const padLength = 16 - (buffer.length % 16);
    const padding = Buffer.alloc(padLength, padLength);
    return Buffer.concat([buffer, padding]);
  }

  pkcs7Unpad(buffer) {
    const padLength = buffer[buffer.length - 1];
    if (padLength > 16 || padLength === 0) throw new Error("Invalid padding");
    return buffer.slice(0, -padLength);
  }
}

// Example usage:
const password = '08180230818209389012832132109123';
const salt = crypto.randomBytes(16);

const aes = new AESCipher(password, salt);

const data = 'This is very secret data!';
const encrypted = aes.encrypt(data);
console.log('Encrypted:', encrypted);

const decrypted = aes.decrypt(encrypted);
console.log('Decrypted:', decrypted);
